pq-crypto/bike/openssl_utils.c

```diff
@@ -0,0 +1,271 @@
+/***************************************************************************
+* Additional implementation of "BIKE: Bit Flipping Key Encapsulation". 
+* Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+*
+* Written by Nir Drucker and Shay Gueron
+* AWS Cryptographic Algorithms Group
+* (ndrucker@amazon.com, gueron@amazon.com)
+*
+* The license is detailed in the file LICENSE.md, and applies to this file.
+* ***************************************************************************/
+
+#include "openssl_utils.h"
+#include "utilities.h"
+#include "openssl/bn.h"
+#include <string.h>
+#include <assert.h>
+
+#define MAX_OPENSSL_INV_TRIALS 1000
+
+_INLINE_ void BN_CTX_cleanup(BN_CTX *ctx)
+{
+   if (ctx) {
+      BN_CTX_end(ctx);
+      BN_CTX_free(ctx);
+   }
+}
+
+DEFINE_POINTER_CLEANUP_FUNC(BN_CTX*, BN_CTX_cleanup);
+
+// Perform a cyclic product by using OpenSSL
+_INLINE_ ret_t ossl_cyclic_product(OUT BIGNUM *r,
+                                   IN const BIGNUM *a,
+                                   IN const BIGNUM *b,
+                                   BN_CTX *bn_ctx) {
+   BIGNUM *m = BN_CTX_get(bn_ctx);
+   if (NULL == m) {
+      BIKE_ERROR(EXTERNAL_LIB_ERROR_OPENSSL);
+   }
+
+   // m = x^PARAM_R - 1
+   if ((BN_set_bit(m, R_BITS) == 0) || (BN_set_bit(m, 0) == 0)) {
+      BIKE_ERROR(EXTERNAL_LIB_ERROR_OPENSSL);
+   }
+
+   // r = a*b mod m
+   if (BN_GF2m_mod_mul(r, a, b, m, bn_ctx) == 0) {
+      BIKE_ERROR(EXTERNAL_LIB_ERROR_OPENSSL);
+   }
+
+   return SUCCESS;
+}
+
+_INLINE_ ret_t invert_poly(OUT BIGNUM *r,
+                           IN const BIGNUM *a,
+                           BN_CTX *bn_ctx) 
+{
+   BIGNUM *m = BN_CTX_get(bn_ctx);
+
+   if (NULL == m) {
+      BIKE_ERROR(EXTERNAL_LIB_ERROR_OPENSSL);
+   }
+
+   // m = x^PARAM_R - 1
+   if ((BN_set_bit(m, R_BITS) == 0) ||
+       (BN_set_bit(m, 0) == 0)) {
+      BIKE_ERROR(EXTERNAL_LIB_ERROR_OPENSSL);
+   }
+
+    // r = a*b mod m
+    uint32_t inv_res;
+    for (uint32_t i = 0; i < MAX_OPENSSL_INV_TRIALS; i++) {

```

Comment:
Maybe I'm misunderstanding something, but why are we calling `BN_GF2m_mod_inv()` in a loop? The documentation for `BN_GF2m_mod_inv()` says "Invert a, reduce modulo p, and store the result in r. r could be a.", and as far as I can tell niether `a`, `m`, or `bn_ctx` is changing so if it fails the first time it should continue failing. Am I understanding this correctly?

Reply:
See full explanation here: https://github.com/open-quantum-safe/openssl/issues/42